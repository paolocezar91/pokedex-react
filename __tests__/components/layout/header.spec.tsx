import PokeNavbar from '@/components/layout/navbar/navbar';
import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import mockRouter from 'next-router-mock';
import { act } from 'react';

jest.mock('next/router', () => jest.requireActual('next-router-mock'));
jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));


describe('Header', () => {
  beforeEach(() => {
    // Reset the router before each test
    mockRouter.setCurrentUrl('/');
  });

  it('should render title', () => {
    const title = 'title';
    render(<PokeNavbar title={title} />);
    expect(screen.queryByRole('heading', {level: 1})).toBeInTheDocument();
  });

  it('should sent to home page', () => {
    const title = 'title';
    render(<PokeNavbar title={title} />);
    const homePageLink = screen.getByTestId("home-page-link");
    fireEvent.click(homePageLink);

    expect(mockRouter.asPath).toBe('/pokedex');
  });

  it('should update input value and submit form', () => {
    const title = 'title';
    render(<PokeNavbar title={title} />);

    // Get form elements
    const form = screen.getByTestId('form-go-to');
    const input = screen.getByPlaceholderText('actions.go.placeholder');

    // Simulate user typing
    // Simulate user typing
    fireEvent.change(input, { target: { value: 'bulbasaur' } });
    expect(input).toHaveValue('bulbasaur');

    // This will work because testing-library's fireEvent creates proper event structure
    fireEvent.submit(form);

    // Verify router navigation
    expect(mockRouter.asPath).toBe('/pokedex/bulbasaur');
  });

  it('should not navigate with empty input', () => {
    const title = 'title';
    render(<PokeNavbar title={title} />);

    const form = screen.getByTestId('form-go-to');
    fireEvent.submit(form);

    // Router should not change
    expect(mockRouter.asPath).toBe('/');
  });

  it('should handle different pokemon names', () => {
    const title = 'title';
    render(<PokeNavbar title={title} />);

    const testCases = [
      { input: '25', expected: '/pokedex/25' },
      { input: 'pikachu', expected: '/pokedex/pikachu' },
      { input: 'CHARIZARD', expected: '/pokedex/charizard' }, // test case insensitivity
      { input: ' eevee ', expected: '/pokedex/eevee' }, // test whitespace
    ];

    testCases.forEach(({ input, expected }) => {
      const form = screen.getByTestId('form-go-to');
      const inputField = screen.getByPlaceholderText('actions.go.placeholder');

      fireEvent.change(inputField, { target: { value: input } });
      fireEvent.submit(form);

      expect(mockRouter.asPath).toBe(expected);

      // Reset router for next test case
      act(() => {
        mockRouter.setCurrentUrl('/');
      });
    });
  });

  it('should match snapshot', () => {
    const title = 'title';
    const { asFragment } = render(<PokeNavbar title={title} />);
    expect(asFragment()).toMatchSnapshot();
  });
});