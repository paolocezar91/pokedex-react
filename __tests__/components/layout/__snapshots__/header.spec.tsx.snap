// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match snapshot 1`] = `
<DocumentFragment>
  <nav
    class="relative navbar flex flex-wrap items-center justify-between border-b-4 border-solid border-(--pokedex-red-light)"
  >
    <div
      class="flex w-full pt-1 pb-1 mx-auto items-center justify-between"
    >
      <button
        aria-description="Go to Home Page"
        class="navbar-brand pt-2 pb-2 flex items-center btn btn-primary"
        data-testid="home-page-link"
        type="button"
      >
        <img
          alt="title"
          class="mx-2 inline"
          data-nimg="1"
          decoding="async"
          height="48"
          src="/logo.svg"
          style="color: transparent;"
          width="48"
        />
        <h1
          class="text-sm hidden md:inline"
        >
          title
        </h1>
      </button>
      <span
        class="go-to border-solid border-2 border-black bg-white rounded mr-4"
      >
        <div
          class="tooltip-container"
        >
          <div
            class="tooltip-trigger"
          >
            <form
              data-testid="form-go-to"
            >
              <input
                class="w-30 ml-2 py-1 text-black bg-white placeholder-gray-500 text-xs"
                name="pokemon-search"
                placeholder="actions.go.placeholder"
                type="text"
              />
              <button
                class="px-2 text-sm text-white bg-(--pokedex-blue) hover:bg-(--pokedex-blue-dark) py-1 border-l-2 border-solid border-black"
                type="submit"
              >
                actions.go.button!
              </button>
            </form>
          </div>
        </div>
      </span>
    </div>
  </nav>
</DocumentFragment>
`;
